Central cp multi upfs

a-jamshed opened the pull request:
The following changes have been made.

1- By default, CP/DP(s) communication bus is ZMQ (``ZMQ_DIRECT``). There is no need to launch ZMQ streamer.
2- DP(s) can now dynamically register with a central CP using a northbound interface in CP. The order of NFs now changes: CP needs to be launched first, followed by DP(s).
3- Documentation has been updated.

This has been tested with `il_trafficgen`` (ZMQ mode), and Spirent (ZMQ as well as UDP modes). Testing with multiple DPs is still pending.

thakurajayL commented:
@ajamshed  You might want to check why CI/CD failed.

ajamshed commented:
> @ajamshed You might want to check why CI/CD failed.

Oops. I guess my last patch did not fix this. I will take a look. Thanks.

thakurajayL commented on config/interface.cfg:
```diff
@@ -48,9 +48,9 @@ fpc_topology_port = 8181
-; These values are unused when DSDN_ODL_BUILD CFLAG not defined
-cp_nb_ip = 192.168.125.60
-cp_nb_port = 9997
+; These values are used by CP to fetch CP's IP address and registration port
```
Is this the same port where CP waits for incoming connections from DP ? Comment looks bit different

ajamshed replied:
Yes. I will make the comment more clear.

ajamshed replied:
Done.

thakurajayL commented on cp/custom-cp.mk:
```diff
@@ -3,4 +3,9 @@
 #ZMQ_COMM flag is set for direct ZMQ communication, unset for direct UDP communication
 #ZMQ communication is enabled by default
-#CFLAGS += -DZMQ_COMM
+CFLAGS += -DZMQ_COMM
```
If possible, We should have a common custom.mk which is referred by cp/dp.

ajamshed replied:
Okay.

krsna1729 replied:
Or should interface be compiled separately with it's own makefile? Have that as dependency for cp and dp?

ajamshed replied:
Good idea. Will work on this.

ajamshed replied:
For the time being, I have created a common file with the name of `vars.mk` that is shared by both CP and DP.

thakurajayL commented on cp/packet_filters.c:
```diff
@@ -766,11 +782,67 @@ parse_adc_rules(void)
+		}
+	}
+
+	sleep(1);
```
May be add comment. Why sleep is required.

ajamshed replied:
Good point. This was done because of the comment mentioned over [here](https://github.com/omec-project/ngic-rtc/blob/2ec6582d9561e393daff5556c556dbbc2bc08c77/cp/packet_filters.c#L652). Let me remove this line and see whether the dataplane still works fine.

ajamshed replied:
Okay. I tried removing this `sleep()` function and I did not see any DP crash (as the comment suggested). I tried running the `master` version without the `sleep()` function and it did not crash as well. This may mean that, in my testbed, I may be unable to recreate reported crash. I may need verification from someone else before I remove this suspiciously redundant line.

thakurajayL replied:
Yeah. I believe you can keep it for now.

ajamshed replied:
Sorry. I missed this comment before. I am not sure why the `sleep()` is really needed. This `sleep()` was still there even before the ACK logic was added (in the master branch). Let me contact others (Somnath etc.) and get back to you on this.

thakurajayL commented on dp/pipeline/epc_packet_framework.c:
```diff
@@ -84,6 +87,37 @@ static void *dp_zmq_thread(__rte_unused void *arg)
+	if (zmq_send(requester, (void *)&zmq_dp_ip, sizeof(zmq_dp_ip), 0) == -1) {
+		rte_exit(EXIT_FAILURE, "Failed to send registration request to CP!\n");
+	}
+	if (zmq_recv(requester, buffer, LEN, 0) == -1) {
```
all zmq calls are blocking ?

ajamshed replied:
ZMQ calls are blocking by default (and are set as such over here).

thakurajayL commented on interface/interface.c:
```diff
@@ -218,6 +270,154 @@ int process_resp_msg(void *buf)
+{
+	static char addr_string[128] = {0};
+	snprintf(addr_string, sizeof(addr_string),
+		 "%s://*:%u", "tcp", cp_nb_port);
```
Any specific reason why we are not using cp_nb_addr also in the bind ?

ajamshed replied:
* represents `INADDR_ANY` over here.

thakurajayL replied:
Yes understood. So control plane is binding on INADDR any and data plane is suppose to send the registration on the cp_nb_ip.  It looks good. But just that control plane is not binding to the specific address as per config. But looks good. I thought you missed it, but looks like you intentionally binding it to *.

ajamshed replied:
Yes.

thakurajayL reviewed (commented):
I am done with code review. May be others can see once.

a-jamshed committed:
Split CP communication across multiple UPFs.

Added logic for dynamic regisration of UPFs.
Logic for dynamic de-registation is not there at the moment.
Added documentation, cleaned up the code, made default ZMQ_COMM as ZMQ_DIRECT.
There is no need to run ZMQ_Streamer now. Added bare minimum documentation.
Dynamic registration of DP(s) should work. Order of NF initializations now changes.
Run CP. Then DP(s).

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

Commit:
SPGWU should support dynamic config for dp selection #90

ajamshed commented on cp/packet_filters.c:
```diff
@@ -766,11 +782,67 @@ parse_adc_rules(void)
 }
+#if defined(CP_BUILD) && defined(MULTI_UPFS)
+void
+init_pkt_filter_for_dp(void)
```
Maybe. I will take a look at this function again once #91 is done.

ajamshed commented on interface/interface.c:
```diff
@@ -218,6 +270,154 @@ int process_resp_msg(void *buf)
+	static const char ok[] = "ACK";
+	static const size_t ok_len = 3; /* strlen(ok) */
+	memset(&addr, 0, sizeof(addr));
+	int n = zmq_recv(dp_sock, &addr, sizeof(addr), 0);
```
I will take a look at this further once #91 is finalized.

a-jamshed committed:
Cleaned up the code.

Added more comments to the code.
Added error handling around sending messages over ZMQ sockets.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

Commit:
SPGWU should support dynamic config for dp selection #90

a-jamshed committed:
Pass S1U's IP address during DP's registration as well.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

krsna1729 commented on interface/interface.c:
```diff
@@ -353,11 +359,16 @@ check_for_new_dps(void)
+	/* DP's IP in the first 32 B; S1U's IP address to the last 32 B */
+	addr = (uint32_t)(msg_bundle >> (sizeof(addr) * CHAR_BIT));
 	a.s_addr = addr;
+	s1u_sgw_ip.s_addr = (msg_bundle & 0xFFFFFFFF);
```
@ajamshed This does not seem to be per DP?

ajamshed replied:
Not yet. This will eventually go in, _e.g._ ``ue_context`` struct once #90 is merged. Currently this patch makes the CP's run-time argument of passing S1U_SGW_IP redundant.

Commit:
SPGWU should support dynamic config for dp selection #90

a-jamshed committed:
Merge branch 'central-cp-multi-upfs' of https://github.com/thakurajayL/ngic-rtc into thakurajayL-central-cp-multi-upfs

a-jamshed committed:
Merge branch 'thakurajayL-central-cp-multi-upfs' into central-cp-multi-upfs

a-jamshed committed:
Merging Ajay's changes.

Minor cleanup to the codebase
Replaced (f)printfs with RTE_LOG_DP's. Added error checkings. Removed a LIST iteration issue.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

a-jamshed committed:
Added multi-upf control (via CP) logic

This commit fixes a few bugs (parsing of app_config.cfg) and also completes the multi-upf control via CP (beta version).
Please note the new contents of app_config.cfg. I tested this code against Spirent LS. And it works without adding quotes
around fields. upf_context as well as dp_info now has additional fields to support the multi-upf control. Also, note the
new paths of CP_CONFIG_ETC_PATH & CP_CONFIG_OPT_PATH. The macro names should be changed.

Next commit will contain updates to documentation and also renaming of a few macros.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

a-jamshed committed:
Added logic so that a DP dynamically re-connects to the CP (in case the DP restarts).

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

thakurajayL committed:
Merge pull request #1 from omec-project/central-cp-multi-upfs

Central cp multi upfs

thakurajayL committed:
Cleaning up multi upf selection based on TAC

thakurajayL committed:
Removal of MME address configuration from the SPGW. SPGW
need not worry about MME address. MME address is picked up
from the CSReq message FTEID IE.

thakurajayL committed:
EDGEPOD-103 : SGW should not take MME address as config parameter

a-jamshed committed:
Fixes to dynamic de-registration.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

a-jamshed committed:
Merge pull request #97 from thakurajayL/central-cp-multi-upfs

Central cp multi upfs

a-jamshed committed:
Code cleanup.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

a-jamshed committed:
Merge pull request #98 from thakurajayL/multiupf-mmeip-dep-removal

Multiupf mmeip dep removal

a-jamshed committed:
Code cleanup.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

thakurajayL committed:
Merge pull request #3 from omec-project/central-cp-multi-upfs

Central cp multi upfs

thakurajayL committed:
EDGEPOD-107 : Control Plane SPGWC crashes if unknown DP connects

If unknown DP connects to control plane then control plane should just
log the message and continue. Not a good idea to crash. Since it leads
to outage of control plane.

thakurajayL committed:
Fixing compilation issue

thakurajayL committed:
Fixing logging macors. Moving them all to single file and removing it
from multiple places.

a-jamshed committed:
Fixed a few error checks and other formatting edits.

Don't do hard exists if the UE tries to attach without any registered UPF.
Fixed an indentation.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

thakurajayL committed:
EDGEPOD-177 - SPGW should support PCO

SPGW to keep config per edge & global default config.
If no dns config in the app_config.cfg file then at code level
default dns are kept as 8.8.8.8 & 8.8.8.4

Now PGW responds to PCO only in CSReq message.

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into central-cp-multi-upfs

Commit:
Add .fossa.yml to central-cp-multi-upfs branch for FOSSA CI

Commit:
Using shorter license header for .fossa.yml

Commit:
Merge pull request #104 from sindack/central-cp-multi-upfs

Add .fossa.yml to central-cp-multi-upfs branch for FOSSA CI

thakurajayL committed:
EDGEPOD-177 - SPGW should support PCO

Code review comment fixes.

thakurajayL committed:
EDGEPOD-212 : configmap update is not updating the config in the spgw

SPGW config update was not getting relected in the running POD.
Corrected config file paths.

a-jamshed committed:
Merge branch 'central-cp-multi-upfs' of https://github.com/thakurajayL/ngic-rtc into thakurajayL-central-cp-multi-upfs

a-jamshed committed:
Merge branch 'thakurajayL-central-cp-multi-upfs' into central-cp-multi-upfs

a-jamshed committed:
Better code indentation.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into config-update-fix

thakurajayL committed:
EDGEPOD-212 : configmap update is not updating the config in the spgw

New base config folder is taken as a argument. THis will keep the
code deployment agnostic

thakurajayL committed:
Merge branch 'config-update-fix' of github.com:thakurajayL/ngic-rtc into config-update-fix

a-jamshed committed:
Merge branch 'thakurajayL-config-update-fix' into central-cp-multi-upfs

Commit:
Static address allocation support by PDN GW

SPGW takes command line argument of static pool if there is signle DP.
If multiple DP then command line argument is ignored and static pool
coming through app_config.cfg file is used.
When CSReq comes with non zero address, then address is validated to be
present in the range. If yes then CSReq is accepted and the same address
is returned back to MME in CSRsp.

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into static-addr

a-jamshed committed:
Minor error checks & better code indentations.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

krsna1729 commented:
retest this please

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into static-addr

thakurajayL committed:
EDGEPOD-189 - Extra MBRsp coming out of SGW

thakurajayL committed:
Merge pull request #4 from omec-project/central-cp-multi-upfs

Central cp multi upfs

thakurajayL committed:
Merge pull request #5 from omec-project/central-cp-multi-upfs

Central cp multi upfs

thakurajayL committed:
EDGEPOD-189 - Extra MBRsp coming out of SGW

Fixing UT issues.

thakurajayL committed:
EDGEPOD-189 - Extra MBRsp coming out of SGW

Removing .orig & .rej files

a-jamshed committed:
Merge branch 'thakurajayL-mbrsp' into central-cp-multi-upfs

a-jamshed committed:
Better indentation.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

thakurajayL committed:
Static address allocation support by PDN GW

Fixing review comments.

thakurajayL committed:
Static address allocation support by PDN GW

Fixing code review comments.

thakurajayL committed:
Static address allocation support by PDN GW

Fixing code review comments.

thakurajayL committed:
Configuration consistency

Moving all the file reference as relative to config base folder.
We can always keep multipe config folders and switch from one
config to other by passing different base directory as argument to
program.

thakurajayL committed:
Static address allocation support by PDN GW

addressing code review comments.

thakurajayL committed:
Static address allocation support by PDN GW

code review comments.

a-jamshed committed:
Merge branch 'static-addr' of https://github.com/thakurajayL/ngic-rtc into thakurajayL-static-addr

a-jamshed committed:
Merge branch 'thakurajayL-static-addr' into central-cp-multi-upfs

a-jamshed committed:
Cosmetic updates to the code.

Fixed a few error check(s) and indentations.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

thakurajayL committed:
Merge pull request #6 from omec-project/central-cp-multi-upfs

Central cp multi upfs

krsna1729 committed:
Revert "Configuration consistency"

This reverts commit 29b094eb7413ecacf5901f4af7d3c5e85ff3c2c8.

thakurajayL committed:
Merge pull request #7 from omec-project/central-cp-multi-upfs

Revert "Configuration consistency"

a-jamshed committed:
Rectified IPv4 verification condition.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>

thakurajayL committed:
EDGEPOD-312 - SPGW to use more ids to support support multiple plmns

Use mcc, mnc & TAC number to select edge

hyunsun committed:
Apply AVX patch to the Docker image

a-jamshed committed:
Merge pull request #120 from ajamshed/central-cp-multi-upfs

Rectified IPv4 verification condition.

a-jamshed committed:
Merge branch 'central-cp-multi-upfs' into central-cp-multi-upfs

a-jamshed committed:
Merge pull request #124 from hyunsun/central-cp-multi-upfs

Apply AVX patch to the Docker image

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into multi-plmn

thakurajayL committed:
EDGEPOD-352 : Send MTU in the PCO (#127)

Send IPv4 mtu in the pco options to UE.

thakurajayL committed:
Merge pull request #10 from omec-project/central-cp-multi-upfs

Central cp multi upfs

yoooou committed:
Fix CI by adding CP northbound server IP and port (#125)

* Fix CI by adding CP northbound server IP and port

* More fixes for CI

thakurajayL committed:
EDGEPOD-356 : SPGWC crashes after config update

While we switch from old config to new config, we should use the previous
upf context, s1u address and static pool trees as is.

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into central-cp-multi-upfs

thakurajayL committed:
EDGEPOD-356 : SPGWC crashes after config update

Fixing review comments

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' of https://github.com/thakurajayL/ngic-rtc into central-cp-multi-upfs

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into multi-plmn

thakurajayL committed:
Test code for checking jenkins issue. Will remove it once issue is resolved.

thakurajayL committed:
Fixing jenkins

thakurajayL committed:
Fixing jenkins issue

thakurajayL committed:
Merge branch 'multi-plmn' of https://github.com/thakurajayL/ngic-rtc into multi-plmn

thakurajayL committed:
Jenkin issue fix.

thakurajayL committed:
Fixing jenkins

a-jamshed committed:
Merge pull request #128 from thakurajayL/central-cp-multi-upfs

Fix: upf context during config update

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into multi-plmn

thakurajayL committed:
fixing compilation issue

thakurajayL committed:
Merge branch 'multi-plmn' of https://github.com/thakurajayL/ngic-rtc into multi-plmn

thakurajayL committed:
fixing jenkins issue

a-jamshed committed:
Merge pull request #122 from thakurajayL/multi-plmn

Multi plmn

thakurajayL committed:
Merge pull request #11 from omec-project/central-cp-multi-upfs

Central cp multi upfs

yoooou committed:
Fix CI by replacing apn1 with internet

Commit:
Merge pull request #129 from yoooou/central-cp-multi-upfs

Fix CI by replacing apn1 with internet

Commit:
enable DDN on dp (#133)

* add config file to ngic-rtc repo for fossa license scan

* Add libzmq5 to rundeps

* delete fossa file to solve the conflict

* enable DDN feature on DP

  DDN feature is developed and tested within EPC. So we can enable it in DP now.

* bug fix for a type defination

Co-authored-by: Hyunsun Moon <hyunsun.moon@gmail.com>

a-jamshed committed:
Fixed endianness orders for a few messages.

Fixed messsages that would be passed between SPGW-C and upf.
Replaced an error statement with something more intelligent string.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>
Signed-off-by: Saikrishna Edupuganti <saikrishna.edupuganti@intel.com>
Signed-off-by: Hynsun Moon <hyunsun@opennetworking.org>

thakurajayL committed:
Merge pull request #13 from omec-project/central-cp-multi-upfs

Central cp multi upfs

a-jamshed committed:
Merge pull request #137 from ajamshed/topics/fix-endian-order

Fixed endianness orders for a few messages.

thakurajayL committed:
Reject CSReq is no user plane found. Also response to keepalive messages.

thakurajayL committed:
Merge branch 'central-cp-multi-upfs' into central-cp-multi-upfs

a-jamshed committed:
Merge pull request #138 from thakurajayL/central-cp-multi-upfs

Central cp multi upfs
